cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringServerHandler.java

```diff
@@ -141,6 +143,52 @@ public void connectControlChannel(FullHttpRequest request, HttpResponder respond
                                    TetheringControlResponse[].class));
   }
 
+  /**
+   * Sends control commands to the client and receives program status updates from the client.
+   */
+  @POST
+  @Path("/tethering/channels/{peer}")
+  public void pollControlChannel(FullHttpRequest request, HttpResponder responder, @PathParam("peer") String peer)
+    throws IOException, NotImplementedException, BadRequestException, TopicNotFoundException {
+    checkTetheringServerEnabled();
+    PeerInfo peerInfo;
+    try {
+      peerInfo = store.getPeer(peer);
+    } catch (PeerNotFoundException e) {
+      responder.sendString(HttpResponseStatus.OK,
+                           GSON.toJson(new TetheringControlResponseV2(Collections.emptyList(),
+                                                                      TetheringStatus.NOT_FOUND)));
+      return;
+    }
+    store.updatePeerTimestamp(peer);
+    TetheringStatus tetheringStatus = peerInfo.getTetheringStatus();
+    if (tetheringStatus != TetheringStatus.ACCEPTED) {
+      // Don't send control messages to a peer that's not in ACCEPTED state.
+      TetheringControlResponseV2 response = new TetheringControlResponseV2(Collections.emptyList(),
+                                                                           tetheringStatus);
+      responder.sendJson(HttpResponseStatus.OK, GSON.toJson(response));
+      return;
+    }
+
+    String messageId = processRequestContent(request, peer);
+    List<TetheringControlMessageWithId> controlMessages;
+    try {
+      controlMessages = getControlMessages(peer, messageId);
+    } catch (TopicNotFoundException e) {
+      // This can only happen if we added the peer during tethering creation, but crashed before creating the topic.
+      // Recreate the topic and pull control messages again.
+      TopicId topic = new TopicId(NamespaceId.SYSTEM.getNamespace(), topicPrefix + peer);
+      createTopicIfNeeded(topic);
+      controlMessages = getControlMessages(peer, messageId);
```

Comment:
we need to catch the TopicNotFoundException and throw some exception that results in a 500, since we don't want this scenario to result in a 404.

Reply:
Discussed offline; TopicNotFoundException will result in a 500 as it's not an instance of NotFoundException